From d28d1f10f99f8ab2fe2bd06c3d22397d0ba08687 Mon Sep 17 00:00:00 2001
From: David Woodhouse <dwmw2@infradead.org>
Date: Thu, 21 Sep 2006 08:55:28 +0100
Subject: [PATCH 10/15] Remove UML header export

From: David Woodhouse <dwmw2@infradead.org>

No need for UML to export headers for userspace to build against.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 include/asm-um/Kbuild |    1 -
 1 file changed, 1 deletion(-)

--- linux-2.6.18.orig/include/asm-um/Kbuild
+++ /dev/null
@@ -1 +0,0 @@
-include include/asm-generic/Kbuild.asm
