From stable-bounces@linux.kernel.org  Sun Apr 17 17:10:52 2005
Date: Sun, 17 Apr 2005 17:05:19 -0700
From: "David S. Miller" <davem@davemloft.net>
To: stable@kernel.org
Subject: [PATCH] sparc64: Fix copy_sigingo_to_user32()

From: jurij@wooyd.org

Because this routine was not filling in the siginfo
values for si_band and si_fd, this broke applications
trying to actually get at this data.

This makes the sparc64 code in line with PowerPC64's
implementation, which already gets it right.

Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Chris Wright <chrisw@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

--- 1.52/arch/sparc64/kernel/signal32.c	2005-03-13 15:29:47 -08:00
+++ edited/arch/sparc64/kernel/signal32.c	2005-04-17 14:37:13 -07:00
@@ -192,9 +192,12 @@
 			err |= __put_user(from->si_uid, &to->si_uid);
 			break;
 		case __SI_FAULT >> 16:
-		case __SI_POLL >> 16:
 			err |= __put_user(from->si_trapno, &to->si_trapno);
 			err |= __put_user((unsigned long)from->si_addr, &to->si_addr);
+			break;
+		case __SI_POLL >> 16:
+			err |= __put_user(from->si_band, &to->si_band);
+			err |= __put_user(from->si_fd, &to->si_fd);
 			break;
 		case __SI_RT >> 16: /* This is not generated by the kernel as of now.  */
 		case __SI_MESGQ >> 16:

